### PR TITLE
Fix CI, add Meson build system, fix CMake, add install feature

### PR DESCRIPTION
**Pull request by scivision**
* corrected errors in Travis-CI and AppVeyor setup
* corrected errors in CMake (never glob source files, must set C++14 std)
* add Meson, a modern alternative to CMake

### CODE REVIEW

**Comment by cawvyoct** [comment]
:wink: 
sweet.

**Comment by cawvyoct** [comment]
:+1:  Replying to say thanks. Initial scan-read seems v.good. 
Will review your PR very soonish ok?

> corrected errors in CMake (never glob source files, must set C++11 std)

Like eating a tasty slice of pizza. :pizza: Hits the spot just right! :wink:

**Comment by cawvyoct** [comment]
Thanks again @scivision, I reviewed your PR.
* I really liked the clean-up to using modern CMake standards. :+1: 
* Also, kudos for the added Meson support! :tada: 
* Great work on updating the travis-ci script! :gear: 
Planning was on our `TODO:` to update the script to use a newer Ubuntu / compiler so you have saved us quite some time with that!
_(Will have to confirm that the script works correctly when merged however...)_

I have noticed that the `CMakeLists.txt` (and the `meson.build` equiv.) now lives in the `src/` directory. 
It would be preferred if the `CMakeLists.txt` was in the "top" / "root" directory of this Github project. 

If possible, could you make a root / top-level `CMakeLists.txt` (and the `meson.build` equiv.) which **_either_**:

* adds to the project's `SOURCES` by _relative-linking_ to the files in the `src/` directory .
(`set(SOURCES src/2048.cpp  src/gameboard.cpp ...` etc)
_N.B: This choice is preferred unless there are objections._
* top-level `CMakeLists.txt` _"includes"_ another `CMakeLists.txt` in the `src/` directory.

A better explanation / example for _Point 2_...:
```
/
CMakeLists.txt # (Main "2048" "PROJECT" file, Imports "src/CMakeLists.txt")
               # (Also has "target_include_directories" information here too)
meson.build(?) # (Does a similar thing as the CMake equiv.)
[...]

---src/
   CMakeLists.txt # (Lists files for SOURCE of 2048 PROJECT)
   meson.build(?)
   2048.cpp
   game.cpp
   [...]

   ---header/
      2048.hpp
      game.hpp
      [...]

README.md
.clang-format
[...]
```

Your changes are greatly appreciated and hope to be seeing more PRs again in the future from you! :cat: :tada:

**Comment by scivision** [comment]
8e04134 moves the build scripts back to the top level directory

**Comment by cawvyoct** [comment]
Great job on the update, thanks. :+1: 

* :warning: Commit 5867861: 
This sets the project's standard to accept `C++14` code. 
The project's C++ standard is `C++11` for now _(educational purposes)_ so please revert.

---

The following comments are mainly style changes for `CMakeLists.txt`
The styles suggested follows the ["CMake Developers Reference Page"][dev-style] mainly the sections:
* [Styles][1]
* [Sample Find Module][2]

[dev-style]: https://cmake.org/cmake/help/latest/manual/cmake-developer.7.html
[1]: https://cmake.org/cmake/help/latest/manual/cmake-developer.7.html#style
[2]: https://cmake.org/cmake/help/latest/manual/cmake-developer.7.html#a-sample-find-module

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,18 +1,34 @@
-cmake_minimum_required(VERSION 3.8 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.7)  # 3.7: CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT, 3.3: CXX_STANDARD
 project(2048 CXX)
 
-set(MAIN_TARGET_NAME 2048)
+set(SOURCES src/2048.cpp src/gameboard.cpp src/game.cpp src/global.cpp src/menu.cpp src/scores.cpp src/statistics.cpp)
```
Comment (cawvyoct) [inline CMakeLists.txt]: This should be:
```cmake
set(SOURCES
  src/2048.cpp
  src/gameboard.cpp
  src/game.cpp
  src/global.cpp
  src/menu.cpp
  src/scores.cpp
  src/statistics.cpp
)
```

**Comment by scivision** [comment]
The code doesn't compile with C++11 in general, for example it will error on Intel compiler, but will compile with C++14.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,18 +1,34 @@
...
+  list(APPEND FLAGS -Wall)
+elseif(CMAKE_CXX_COMPILER_ID STREQUAL Intel)
+  list(APPEND FLAGS -w2)
+elseif(CMAKE_CXX_COMPILER_ID STREQUAL PGI)
+  list(APPEND FLAGS -a)
+endif()
 
-add_executable(${MAIN_TARGET_NAME} ${SOURCES} ${HEADERS})
+add_executable(2048 ${SOURCES})
+target_include_directories(2048 PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/src/headers/)
```
Comment (cawvyoct) [inline CMakeLists.txt]: This should be:
```cmake
target_include_directories(2048 PRIVATE
  ${CMAKE_CURRENT_SOURCE_DIR}/src/headers/
)
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,18 +1,34 @@
...
+target_compile_options(2048 PRIVATE ${FLAGS})
```
Comment (cawvyoct) [inline CMakeLists.txt]:
```cmake
target_compile_options(2048 PRIVATE
  ${FLAGS}
)
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,18 +1,34 @@
...
 
-target_compile_features(${MAIN_TARGET_NAME} PRIVATE cxx_std_11)
+set_target_properties(2048 PROPERTIES
+  CXX_STANDARD 14
+  CXX_STANDARD_REQUIRED true)
```
Comment (cawvyoct) [inline CMakeLists.txt]:
```cmake
set_target_properties(2048 PROPERTIES
  CXX_STANDARD 14
  CXX_STANDARD_REQUIRED true
)
```

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,18 +1,34 @@
...
 
-target_include_directories(${MAIN_TARGET_NAME} PRIVATE "src/headers/")
+# --- install
+
+if(UNIX AND CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
+  set(CMAKE_INSTALL_PREFIX "~/.local" CACHE PATH "..." FORCE)
+endif()
+
+install(TARGETS 2048
+  RUNTIME DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
```
Comment (cawvyoct) [inline CMakeLists.txt]:
```cmake
install(TARGETS 2048
  RUNTIME
    DESTINATION ${CMAKE_INSTALL_PREFIX}/bin
)
```

**Comment by scivision** [comment]
Are there auto-corrections for this listing style? As a CMake contributor, is it necessary to be more fussy that Kitware itself on indents?

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,18 +1,34 @@
...
+
+install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/data
+        DESTINATION ${CMAKE_INSTALL_PREFIX})
```
Comment (cawvyoct) [inline CMakeLists.txt]:
```cmake
install(
  DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/data
  DESTINATION ${CMAKE_INSTALL_PREFIX}
)
```

**Comment by cawvyoct** [comment]
> The code doesn't compile with C++11 in general, for example it will error on Intel compiler, but will compile with C++14.

Is there a particular part of the code which is written in a C++14 format that is erroring-out on the Intel compiler? :mag: 

---

In either case, personally, I am happy to move things to C++14 but again, this is a _"newish? learner"_ project. :mortar_board: 

Quite a few contributors understand _"C++11"_ code (as they have learnt from _"C++03"_ roots), but _"C++14"_-style (and beyond), although is very beneficial to the codebase, is more of a _"challenge"_ to understand.

I believe that newer learners should learn C++14 and beyond _(especially because the world needs more `return auto (T)` and `constexpr` help and love!)_ :cat: but I will leave the final call with @plibither8.

**Comment by plibither8** [comment]
Hi @scivision @cawvyoct :wave:,

Due to inexperience in dealing with CMake directly, I cannot comment on the specifics of CMake, apologies.

Now as for upgrading the project's standard from C++11 to C++14, I'm more than happy about it :smile:. I guess, if it is a _"challenge"_, we should gracefully accept it!

So from my side, there is no hesitation in moving to C++14.

**Comment by plibither8** [comment]
Moreover, I think we should mention in the Readme's [Contributing](https://github.com/plibither8/2048.cpp#contributing) section regarding the standards of both the source code (i.e. C++11/14) and the CMakeLists file.

**Comment by cawvyoct** [comment]
> Moreover, I think we should mention in the Readme's [Contributing](https://github.com/plibither8/2048.cpp#contributing) section regarding the standards of both the source code (i.e. C++11/14) and the CMakeLists file.

Ya, seems most prudent. :cat:

**Comment by cawvyoct** [comment]
@scivision, your PR is good for merging. :cat:  :incoming_envelope:

* :information_source: Would you object to myself doing the stylistic changes quite soon after accepting your PR?

  As we will add the KDE / CMake Dev recommendations to the [Contributing][1] section, to follow what is written on the pages, we will require the changes _(to prevent less admin headache later)_.

  _The changes in **no way indicate an inability to produce any valid conforming PR yourself**_. :dove: 

In this PR's case, as there is nothing broken / malfunctioning within your PR, personally I think it is fine for being accepted.

Once again, thank you for your PR / efforts and we are looking forward to future PRs from you. :+1: 

---
The PR will be closed via _"Squash and Merge"_. :card_file_box: 
_(Will be automatically merged in 12hrs from this message if no objections)_

[1]: https://github.com/plibither8/2048.cpp#contributing

**Comment by scivision** [comment]
I did not investigate the proper resolve for these errors, just noted them.

This is what PGI 18.10 `pgc++` errors on unless C++14 is set:
```
"~/code_other/2048.cpp/src/gameboard.cpp", line 86: error: no operator
          "%" matches these operands
            operand types are: std::initializer_list<int> % int
          point2D_t{index_counter % getPlaySize(), index_counter / getPlaySize()};
                                  ^

"~/code_other/2048.cpp/src/gameboard.cpp", line 86: error: no operator
          "/" matches these operands
            operand types are: std::initializer_list<int> / int
          point2D_t{index_counter % getPlaySize(), index_counter / getPlaySize()};
                                                                 ^

"~/code_other/2048.cpp/src/gameboard.cpp", line 90: error: no operator
          "++" matches these operands
            operand types are: std::initializer_list<int> ++
      index_counter++;
                   ^

"~/code_other/2048.cpp/src/gameboard.cpp", line 144: error: no
          operator "%" matches these operands
            operand types are: std::initializer_list<int> % int
          point2D_t{index_counter % getPlaySize(), index_counter / getPlaySize()};
                                  ^

"~/code_other/2048.cpp/src/gameboard.cpp", line 144: error: no
          operator "/" matches these operands
            operand types are: std::initializer_list<int> / int
          point2D_t{index_counter % getPlaySize(), index_counter / getPlaySize()};
                                                                 ^

"~/code_other/2048.cpp/src/gameboard.cpp", line 145: error: no
          operator "++" matches these operands
            operand types are: std::initializer_list<int> ++
      index_counter++;
                   ^
```

Similarly, Intel 19 `icpc` errors for C++11:
```
~/code_other/2048.cpp/src/gameboard.cpp(86): error: no operator "%" matches these operands
            operand types are: std::initializer_list<int> % int
          point2D_t{index_counter % getPlaySize(), index_counter / getPlaySize()};
                                  ^

~/code_other/2048.cpp/src/gameboard.cpp(86): error: no operator "/" matches these operands
            operand types are: std::initializer_list<int> / int
          point2D_t{index_counter % getPlaySize(), index_counter / getPlaySize()};
                                                                 ^

~/code_other/2048.cpp/src/gameboard.cpp(90): error: no operator "++" matches these operands
            operand types are: std::initializer_list<int> ++
      index_counter++;
                   ^

~/code_other/2048.cpp/src/gameboard.cpp(144): error: no operator "%" matches these operands
            operand types are: std::initializer_list<int> % int
          point2D_t{index_counter % getPlaySize(), index_counter / getPlaySize()};
                                  ^

~/code_other/2048.cpp/src/gameboard.cpp(144): error: no operator "/" matches these operands
            operand types are: std::initializer_list<int> / int
          point2D_t{index_counter % getPlaySize(), index_counter / getPlaySize()};
                                                                 ^

~/code_other/2048.cpp/src/gameboard.cpp(145): error: no operator "++" matches these operands
            operand types are: std::initializer_list<int> ++
      index_counter++;
                   ^
```

**Comment by cawvyoct** [comment]
Thanks for the report. :tada: 
Hummm... according to [cppreference][1], this problem was retroactively fixed to the C++11 standards.

The error is _"std::initializer_list constructors had priority over copy constructors for single-element lists"_.

[Example.][2]
(Try it out in different compilers!
_@plibither8, how would one fix this to work with `icpc++` and `pgc++`?_ :mag: :cat: )

It seems that the erroring-out compilers  (`icpc++`, `pgc++`) have not retroactively done so yet...

_This was also discussed (before the retroactive correction) in [Herb Sutter's GotW 92][3]_

[1]: https://en.cppreference.com/w/cpp/language/list_initialization#Defect_reports
[2]: https://godbolt.org/z/jppqfF
[3]: https://herbsutter.com/2013/06/07/gotw-92-solution-auto-variables-part-1/

**Pull request merged**